### ISSUE TITLE
[Hacking] ChipNBTSet packet

### ISSUE THREAD

**Issue by TheAndrey** [MrTJP/ProjectRed#1384]
I found unsafe packet handler.  
https://github.com/MrTJP/ProjectRed/blob/master/src/mrtjp/projectred/transportation/packethandlers.scala#L178-L179
```scala
    private def setChipNBT(packet:PacketCustom, player:EntityPlayerMP)
    {
        val slot = packet.readUByte()
        val stack = packet.readItemStack()
        player.inventory.setInventorySlotContents(slot, stack)
        player.inventory.markDirty()
}
```
This puts received from client item to player inventory. Hackers can use this to get ANY item on the server!
You can never trust data that you receive from a client. You should always check what you have received from the client.
In this case, you only need to transfer the NBT tag, but not the entire item.

**Comment by Zaxar163** [comment]
@TheAndrey, I can say that developer doesn`t think about servers, because I finded this bug an year ago, but then I join into discord i was kicked.

**Comment by covers1624** [comment]
ProjectRed doesn't have a discord, I've added this to my list and will check it out this weekend.

**Comment by man-of-stone50** [comment]
This is becoming a serious issue on servers.

**Comment by FoxMcloud5655** [comment]
Agreed.  I patched it manually because we couldn't wait for a patch.

**Comment by man-of-stone50** [comment]
I apparently might have a fix just have no idea of how to use it
The current code
https://github.com/MrTJP/ProjectRed/blob/master/src/mrtjp/projectred/transportation/packethandlers.scala#L178-L179
GitHub
MrTJP/ProjectRed
ProjectRed - Extending redstone functionality to FMP

Lines 178 and 179
It can be fixed by replacing line's 175 below with this code
private def setChipNBT(packet:PacketCustom, player:EntityPlayerMP)
    {
        val slot = packet.readUByte()
        val stack = packet.readItemStack()
        val playerstack = player.inventory.getStackInSlot(slot)
        if (stack.getUnlocalizedName == playerstack.getUnlocalizedName)
        {
            player.inventory.setInventorySlotContents(slot, stack)
            player.inventory.markDirty()
        }
        else
        {
            player.addChatMessage(new ChatComponentText("YOU DIRTY HACKER"))
        }
    }

**Comment by TheAndrey** [comment]
@man-of-stone50 this fix is not safe. Hackers can write any NBT tag to item (like item name, lore, enchantments).

**Comment by man-of-stone50** [comment]
Oh damn that sounds bad, we might have to remove ProjectRed Transportation temporarly.

**Comment by TheAndrey** [comment]
@MrTJP Will the exploit be fixed?

**Comment by MrTJP** [comment]
Fixed via 121cca149d190daea1e52c05029dad4dabf96c91.
